
## P4DS Summative Assignment 1

# Data and Algorithms: Cities and Earthquakes

## An exercise in geographic data analysis

Last Modified: 01/02/2023

## Question 1: World Cities

In this coursework exercise, you will download data from the provided link and read it in as a `CSV` file using the Pandas data analysis package for Python.
The data we will use contains a variety of information about cities from around the world.

### Coding Techniques for Working with `DataFrames`

To complete these tasks you will need to access and filter a `DataFrame`. 
The `DataFrame` data structure has many convenient features for extracting and ordering information. Although conceptually it can be thought of as a comptuational
represention of a table, it is quite a complex data structure and takes a while
to master. The following questions can be done with only a small but powerful
set of `DataFrame` operations; and the following examples of typical forms of programming with `DataFrame`s should be useful 
for coding your answers.

#### Loading data from a CSV file into a `DataFrame`

DataFrames are specifically designed to handle data organised in a tabular format.
Hence, as we would expect, since `CSV` is the standard format for tabular data, it is very easy to create a `DataFrame` by loading data from a `CSV` file.

* ```pandas.read_csv(source)``` ---
The  [read_csv function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html#pandas.read_csv) can accept a filename or URL as an argument.

#### Getting the Data for this Question

Download the data file 'worldcities.csv', and put
the file in the same directory as this Jupyter notebook file.
Then, by running the following cell, we can set the 
global variable ```WC_DF``` to a DataFrame containing the information from ```worldcities.csv```. 

In [ ]:
#Please do not run this cell
from A1_Earthquakes_tests import do_tests, do_all_tests, version

version()

In [1]:
## WC_DF Initialisation

import pandas  ## This is the module for creating and manipulating DataFrames

WC_DF = pandas.read_csv("worldcities.csv")

**Be sure to keep the same variable name** `WC_DF` **for this global variable, otherwise most of the
following code will not work and you will break the autograder.**

#### Checking the contents of a DataFrame

Pandas provides the following useful methods that enable you to quickly check the contents of a `DataFrame`:

* ```df.head()``` ---
  For a DataFrame object, ```df```, this method extracts the first 5 rows of data, so you can easily check what the data looks like.
  
* ```df.describe()``` --- for a DataFrame, ```df```, this method provides a table giving an overview of some basic statistical properties of the DataFrame.

Note that the ```head()``` and ```describe()``` methods are actually operations that 
return a new DataFrame object. If this value is returned by the last line of a cell
it will be displayed as a table, but if it is generated elsewhere in the code
you will not see any output unless you use the ```display``` function from the
```IPython.display``` module.

#### Accessing DataFrame columns and rows

Each column of a `DataFrame` is a list-like object called a
`Series`. Elements, and slices of a `Series` can then be accessed in similar
fashion to a list. The following illustrates how to get the `Series` containing
the first 5 elements of the `city` column of `WC_DF`:

In [2]:
top_5_cities = WC_DF["city"][:5]   ## selects the first 5 items of the "city" column.
top_5_cities

0          Tokyo
1       New York
2    Mexico City
3         Mumbai
4      São Paulo
Name: city, dtype: object

In the above output, the left hand column of the displayed value of `top_5_cities` 
shows the index label of each element. One of the differences between a `Series` and an ordinary list is that, whereas a list always has integers for its index labels, a `Series` can have different kinds of values for these. For instance (though there
is no reason to do this for the current assignment) we could set the index values to alphabetic letters, as follows:

In [3]:
top_5_cities.index = list("abcde")
top_5_cities

a          Tokyo
b       New York
c    Mexico City
d         Mumbai
e      São Paulo
Name: city, dtype: object

In [4]:
top_5_cities.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

You can also use ```.values``` to return an `array` of the column values without
the index:

In [5]:
WC_DF["city"][:5].values

array(['Tokyo', 'New York', 'Mexico City', 'Mumbai', 'São Paulo'],
      dtype=object)

An `array` is also a list-like datastructure. It does not have an index. The main difference between a list and an `array` is that the array is optimised for
storing large amounts of information and for efficiently applying numerical and
other operations to all elements of the array. Hence, arrays are usually preferred
to lists when handling large amounts of information, or when storing numerical
vectors.

You can also easily find the column names of the DataFrame using ```.columns```, for example:

In [6]:
WC_DF.columns

Index(['city', 'city_ascii', 'lat', 'lng', 'country', 'iso2', 'iso3',
       'admin_name', 'capital', 'population', 'id'],
      dtype='object')

__Note:__ The `Index` returned here is yet another type of list-like, object. It is similar to an array,
except that it is used for indexing a `Series` or `DataFrame`. You do not usually
need to create or deal with `Index` objects directly, since this is done automatically when you create and minipulate `DataFrame`s. So you will normally only see one, when
you want to look at the columns or rows of a `DataFrame`. But what you should be
aware of, when dealing with `DataFrames`, is that the word _index_ can refer to
several different types of things.

In many cases you can treat `Series`, `array` and `Index` objects like lists and if you want to change them to an ordinary list you can just use the `list` operator,
as in the following:

In [7]:
list(WC_DF.columns)

['city',
 'city_ascii',
 'lat',
 'lng',
 'country',
 'iso2',
 'iso3',
 'admin_name',
 'capital',
 'population',
 'id']

We can refer to rows of a `DataFrame` either by the expression `DF.loc[label]`, where `label` is the index label of the row we want, or by `DF.iloc[n]`,
where `n` is an `int` giving the position of the row in the `DataFrame`.
In the case of `WC_DF`, the labels are integers, so we would get the same result using either. You could test this. You could also see the difference if you try finding a row of `top_5_cities` `DataFrame` defined above, after its index labels have been replaced by letters. In this case you could access rows either using letters, using `loc`, or by `int`s, using `iloc`.

In [8]:
WC_DF.iloc[1]

city               New York
city_ascii         New York
lat                 40.6943
lng                -73.9249
country       United States
iso2                     US
iso3                    USA
admin_name         New York
capital                 NaN
population       19354922.0
id               1840034016
Name: 1, dtype: object

#### Iterrating through the rows of a DataFrame
A convenient way of going through the rows of a `DataFrame` to perform some operation i by using the `iterrows` method in a `for` loop. This enables you to get both the index label and the row itself, for each successive row of the `DataFrame`. The following code is a simple example:

In [9]:
for i, row in WC_DF.iterrows():
    print(i, row['city_ascii'], row['lat'], row['lng'])
    if i >3: break

0 Tokyo 35.685 139.7514
1 New York 40.6943 -73.9249
2 Mexico City 19.4424 -99.131
3 Mumbai 19.017 72.857
4 Sao Paulo -23.5587 -46.625


#### Sorting the rows of a DataFrame

It is easy, and often very useful, to sort the DataFrame by column values using ```.sort_values```, for example:

In [10]:
WC_DF.sort_values(by=["country"], ascending=True)[:10] # Sorts countries by alphabet

,city,city_ascii,lat,lng,country,iso2,iso3,admin_name,capital,population,id
10235,Karukh,Karukh,34.4868,62.5918,Afghanistan,AF,AFG,Herāt,minor,17484.0,1004546127
8527,Kōṯah-ye ‘As̲h̲rō,Kotah-ye `Ashro,34.4500,68.8000,Afghanistan,AF,AFG,Wardak,NaN,35008.0,1004450357
3341,Shibirghān,Shibirghan,36.6580,65.7383,Afghanistan,AF,AFG,Jowzjān,admin,93241.0,1004805783
6050,Khōst,Khost,33.3395,69.9204,Afghanistan,AF,AFG,Khōst,admin,NaN,1004919977
5141,Maḩmūd-e Rāqī,Mahmud-e Raqi,35.0167,69.3333,Afghanistan,AF,AFG,Kāpīsā,admin,7407.0,1004151943
2088,Lashkar Gāh,Lashkar Gah,31.5830,64.3600,Afghanistan,AF,AFG,Helmand,admin,201546.0,1004765445
3210,Gardēz,Gardez,33.6001,69.2146,Afghanistan,AF,AFG,Paktiyā,admin,103601.0,1004468894
6249,Maīdān Shahr,Maidan Shahr,34.3956,68.8662,Afghanistan,AF,AFG,Wardak,admin,NaN,1004798735
2105,Maīmanah,Maimanah,35.9302,64.7701,Afghanistan,AF,AFG,Fāryāb,admin,199795.0,1004622920
7399,Andkhōy,Andkhoy,36.9317,65.1015,Afghanistan,AF,AFG,Fāryāb,minor,71730.0,1004472345


##### Note on __encodings__ of the city name
There are two columns that hold the city name. The first column name is `'city'` and
the second is `'city_ascii'`. There are various different ways in which textual
information can be encoded into bytes. These days [Unicode characters](https://home.unicode.org/)
encoded using [UTF-8](https://en.wikipedia.org/wiki/UTF-8) are pretty standard.
But the older [ASCII](https://en.wikipedia.org/wiki/ASCII) code, which uses
a single byte per character is still commonly used. Unicode provides a huge
variaty of text characters and other symbols, whereas ASCII is quite 
limited (mainly to characters and symbols found in standard English). 
But ASCII and is simpler and in 
some ways easier to deal with than UTF-8. In the following questions you
will be asked to use the ASCII version of the city name (from the `city_ascii` column). This mainly just to make you aware that there are different encodings
of text strings, but it will also prevent cerain problems that could occur in the
Autograder, if different people used different encodings.

#### Filtering DataFrames
By _filtering_ we mean keeping some parts that we want and throwing away others.
Typically, we look for rows that match some condition; and the filter condition
is often some constraint involving the values for that row in one or more
columns.
`Pandas` `DataFrames` can  be filtered according to the values of a column by using a boolean expression, for example:

In [11]:
filtered_DF = WC_DF[ WC_DF['capital'] == 'admin'] # This keeps only administrative capitals
filtered_DF.head()

,city,city_ascii,lat,lng,country,iso2,iso3,admin_name,capital,population,id
3,Mumbai,Mumbai,19.0170,72.8570,India,IN,IND,Mahārāshtra,admin,18978000.0,1356226629
4,São Paulo,Sao Paulo,-23.5587,-46.6250,Brazil,BR,BRA,São Paulo,admin,18845000.0,1076532519
5,Delhi,Delhi,28.6700,77.2300,India,IN,IND,Delhi,admin,15926000.0,1356872604
6,Shanghai,Shanghai,31.2165,121.4365,China,CN,CHN,Shanghai,admin,14987000.0,1156073548
7,Kolkata,Kolkata,22.4950,88.3247,India,IN,IND,West Bengal,admin,14787000.0,1356060520


This way of filtering is a very powerful and useful aspect of `DataFrames`. 
However, the
syntax of the filter operation is rather unusual and a bit difficult to understand.

What is happening can be explained by these steps in the way a filter expression
is evaluated:
*  `DF['label']` (where `DF` is any `DataFrame`), gives a `Series` corresponding
    to    the `'label'` column of `DF.
    
*  `a_series == val` is a special use of `==`. When a Boolean operator
    (such `==`, `<` etc.) is applied to a `Series` object
   the result is actually a `Series` of Boolean values (not a single Boolean).
   The new `Series` obtained will have the value `True` for each element where
   the original `Series` satisfies `element == val`, and `False` for the rest.
   
* `DF[ bool_series ]`, is a special kind of slice-like operation, where a boolean
   series is given as a selection argument to the `DataFrame`. It will return 
   a new DF, containing all the rows of `DF` for which `bool_series` has the 
   value `True`. (These rows can be quickly found because the `DataFrame` and
   the Boolean series both have the same `Index`.)
   
You do not necessarily need to follow all of that precise description of filtering
but it will be extremely helpful if you are able to construct filtering
operations similar to the above example. You will see another example below,
in relation to the earthquakes data you will be processing.

### Overview of Question 1 tasks

This question requires you to write functions to carry out the following specific tasks. 

* __Question 1a__ --- Find all city names that include *non-ASCII* characters. __[1 Mark]__

* __Question 1b__ --- Find city names that occur multiple times in the dataset. __[2 Marks]__
* __Question 1c__ --- Create a dictionary of the number of cities in each country that are included in the dataset. __[1 Mark]__

* __Question 1d__ --- Create a `DataFrame` of the largest cities (by population) in the world. __[2 Marks]__
* __Quesiton 1e__  --- Find all cities in a given country whose population is above a given number. __[2 Marks]__

* __Question 1f__ --- Find the total population of people living in cities in a given country. __[2 Marks]__

Full details for each task are explained below.

### Question 1a

As we have seen, the names of some cities include accents or symbols that are not represented in the standard ASCII character set. Write a function `non_ascii_cities`, which returns a `Set` containing all the cities that occur in the `world_cities.csv` dataset whose name **cannot** be properly represented using only ASCII characters. 

Your function should not have any arguments. You should assume that the global variable WC_DF has already been initialised by running the first code cell in this notebook (see above).

In [12]:
# Question 1a answer cell

def non_ascii_cities():
    df = WC_DF.copy()
    non_ascii_cities = set(df[df['city']!=df['city_ascii']]['city'])
    
    return non_ascii_cities
    # Modify to return a set of all non-ascii city names in the world_cities data

In [ ]:
#Please do not run this cell
do_tests( non_ascii_cities )


### Question 1b

One issue that you will discover if you investigate the `worldcities.csv` data is that there are many different cities that have the same name. You may find it interesting to discover which are the most common city names. But for this question you need to write a function `num_cities_occurring_n_times(n)`, such that:
    
* The argument `n` is an integer;
* The return value is an integer of type `int`, which is the number of different city names that occur `n` times in the `worldcities` dataset. 
* The function should consider the city name to be the value in the `city` column. In other words the name in the form that may contain non-ASCII characters.

In [13]:
# Question 1b answer cell

def num_cities_occurring_n_times(n):
    df = WC_DF.copy()
    city_list = {}
    for city in df['city']:
        if city in city_list.keys():
            city_list[city] += 1
        else:
            city_list[city] = 1
    
    cities = [k for k,v in city_list.items() if v ==n]
    
    return int(len(cities))
    # Modify to return a value according to the specification given above

In [ ]:
#Please do not run this cell
do_tests( num_cities_occurring_n_times )

### Question 1c

Write a function that returns a dictionary (a `dict` object), whose keys are all the country name strings that occur in the `worldcities` data and whose values are `int`s giving the number of cities of that country that are included in the dataset.



In [14]:
# Question 1c answer cell
def country_num_cities_dict():
    df = WC_DF.copy()
    country_list = {}
    for country in df['country']:
        if country in country_list.keys():
            country_list[country] += 1
        else:
            country_list[country] = 1
    return country_list

In [ ]:
#Please do not run this cell
do_tests( country_num_cities_dict )

### Question 1d
Write a function `largest_cities_dataframe` that takes an `int` argument `n`  and uses the
pandas DataFrame `WC_DF` to return a new `DataFrame` containing `n` rows corresponding to
the `n` largest cities in terms of population size, in order of decreasing population size.

You should return a dataframe such that it has the same columns as the `WC_DF` and each
row has the same values as a corresponding row of `WC_DF`. It does not matter if the row
indexes are the same. (They may or may not be the same depending on the specific way
that you create the new DataFrame.)

In [15]:
# Question 1d answer cell

def largest_cities_dataframe(n):
    df = WC_DF.copy()
    df.sort_values(by=['population'], ascending=False, inplace=True)
    
    return df.head(n)
    # Modify to return a list of the n cities with the largest population

In [ ]:
#Please do not run this cell
do_tests( largest_cities_dataframe )

__NOTE:__ In answering __1d__ you may assume that no two cities have exactly the same population, which is almost but not quite certain, when dealing with large numbers like this. But, of course, when dealing with quantites where multiple data records could have the same value, we need to be careful, because this may not be the case.
For example, if we are interested in what equipement students own, we might think it would be informative
to find 'the top 10 students owning the most laptops'. In this case there could be: 1 student with 3 laptops, 23 students with 2 laptops, 160 with 1 laptop and 3 who do not own a laptop. In such a case it is not meaningful to pick the 'top 10' in terms of laptop ownership. A similar problem could potentially occur with the earthquakes data that we will look at later, because the earthquake magnitudes are only recorded to 1 decimal place.

### Question 1e
Define a function `big_cities_in_country( country, population)` that takes as arguments a string corresponding to the name of a country and an integer, which referes to a population number.
The function should return a `list` of the form 
<center>
    <tt>[("city1", pop1), ("city2", pop2),... ]</tt>, 
</center>

where each pair
`("cityN", popN)` is a `tuple` consisting of the name, **in ASCII form**, of a city in the given `country`, followed by an `int`, which is the population of that city (according to the worldcities data). The list should include all and only those cities in the country whose population is greater than or equal to the given `popuplation` argument. The list should be ordered so that the `("cityN", popN)` items occur in **increasing** order of the population size `popN`.



In [16]:
# Complete question 1e in this cell

def big_cities_in_country(country, population): # country is a string argument
    
    df = WC_DF.copy()
    df = df[df['country']=='Japan'][['country', 'population', 'city_ascii']].dropna()
    df.sort_values(by='population', ascending=True, inplace=True)

    city_pop = []

    for i,row in df.iterrows():
        city_pop.append((row['city_ascii'], int(row['population'])))
    
    return city_pop # Edit this function to return a list, as specified above

In [ ]:
#Please do not run this cell
do_tests( big_cities_in_country )

### Question 1f
Create a function that given a country name, returns an `int` which is the total population of 
people liveing in all the cities of that country, as given in `WC_DF`. 

**Hints:** 
* You can ignore cities for which no population value is recorded in `WC_DF`. You will need to be able to deal with `NaN` values. A distinctive feature of `NaN` values is that they have been defined so that `x != x` has the value `True` if `x` has a `NaN` value. 

In [17]:
## Question 1f Answer Code Cell

def country_total_cities_population(country):
    df = WC_DF.copy()
    df = df[df['country']==country][['country', 'population', 'city']].dropna()
        
    return df['population'].sum()

In [ ]:
#Please do not run this cell
do_tests( country_total_cities_population )

### Question 2: Earthquakes - Web Access and Pandas DataFrames

It is possible to download live information from the web and process it using the Pandas data analysis package for Python. The data we will use as an example is from the 
[United States Geological Survey (USGS)](https://www.usgs.gov/), which 
provides a wide range of geographic and geological information and data.
We shall be using their data relating to seismological events (i.e. Earthquakes) from around the world, which is published in the form of continually updated CSV files. Information about these feeds can be found [here](https://earthquake.usgs.gov/earthquakes/feed/). 

Questions Overview

* __Q2a__ --- Find earthquakes of a given magnitude or higher.           __[2 mark]__
* __Q2b__ --- Make a DataFrame showing the most powerful quakes          __[2 marks]__
* __Q2c__ --- Make a DataFrame showing distance of quakes from a given location __[3 marks]__
* __Q2d__ --- Identify all cities endangered by earthquakes   __[3 marks]__

### Read in data file
__Note:__ For this question you  __need__ to download and save the file 'all_day.csv'. More information about this and other earthquake feeds provided by USGS can be found [here](https://earthquake.usgs.gov/earthquakes/feed/v1.0/csv.php).

In [18]:
import pandas    ## This is the module for creating and manupulating DataFrames

QUAKE_DF = pandas.read_csv( "all_day.csv" )

#### You can use the following cell to test if you have read the quake data into `QUAKE_DF`

In [19]:
## If QUAKE_DF is a DataFrame, show the first 5 rows
try:
    if type(QUAKE_DF) == pandas.DataFrame:
        display(QUAKE_DF.head())
    else:
        print("QUAKE_DF is not a DataFrame")
except:
    print("QUAKE_DF has not been assigned a value")

,time,latitude,longitude,depth,mag,magType,nst,gap,dmin,rms,...,updated,place,type,horizontalError,depthError,magError,magNst,status,locationSource,magSource
0,2020-11-24T14:32:03.660Z,38.093800,-117.521200,6.20,2.2,ml,12.0,301.36,0.19600,0.08,...,2020-11-24T15:02:38.035Z,"25 km W of Tonopah, Nevada",earthquake,NaN,10.70,NaN,NaN,automatic,nn,nn
1,2020-11-24T14:29:48.800Z,38.803001,-122.772331,0.76,0.4,md,10.0,76.00,0.01014,0.03,...,2020-11-24T14:44:05.559Z,"3km NNW of The Geysers, CA",earthquake,0.3,0.72,0.160,2.0,automatic,nc,nc
2,2020-11-24T14:29:38.517Z,60.563200,-151.575600,58.20,2.0,ml,NaN,NaN,NaN,0.69,...,2020-11-24T14:32:49.625Z,"15 km WSW of Salamatof, Alaska",earthquake,NaN,1.10,NaN,NaN,automatic,ak,ak
3,2020-11-24T14:28:38.100Z,38.175000,-117.872200,6.60,1.2,ml,13.0,108.70,0.01500,0.09,...,2020-11-24T14:33:18.030Z,"31 km SE of Mina, Nevada",earthquake,NaN,0.80,NaN,NaN,automatic,nn,nn
4,2020-11-24T14:07:34.177Z,-22.197700,-176.098100,57.45,5.3,mb,NaN,83.00,6.56000,0.80,...,2020-11-24T14:37:05.040Z,"146 km SW of Vaini, Tonga",earthquake,10.0,6.00,0.065,78.0,reviewed,us,us


##### Note:
The columns containing latitude and longitude values are labelled differently in the `worldcities.csv` and the earthquake data from USGS. This is a minor but very typical form of incompatibility between data formats that you will often need to deal with when working with real data.

### More examples of useful `pandas` functions

Here we show you some more pandas functions that you may find useful in this exercise. 

As we have seen, versatile filtering and sorting capabilities are provided by pandas. To get more understanding of these, you should look at tutorials of using Pandas DataFrames. But the following example illustrates how you can find and display quakes whose depth is greater than or equal to a given threshold:

In [20]:
def show_deep_quakes( depth ):
    # make deep_quakes DataFrame by selecting rows from QUAKE_DF
    deep_quakes = QUAKE_DF[ QUAKE_DF["depth"] >= depth ]  ## This is how you select rows by a condition
                                                          ## on one of the column values.
        
    print("Number of quakes of depth {} or deeper:".format(depth), 
           len(deep_quakes.index))     ## This finds the number of rows of the deep_quakes DataFrame
    
    display(deep_quakes.sort_values("depth", ascending=False))  ## Sort by descending depth value

**Note:**
The `QUAKES_DF` global variable needs to be set before these examples will work, so I am using a `try`, `except` construct to avoid getting an error.

In [21]:
try:
    show_deep_quakes(100)
except:
    print("Probably QUAKE_DF not correctly set")

Number of quakes of depth 100 or deeper: 6


,time,latitude,longitude,depth,mag,magType,nst,gap,dmin,rms,...,updated,place,type,horizontalError,depthError,magError,magNst,status,locationSource,magSource
137,2020-11-23T22:07:51.632Z,-22.6121,-66.1560,255.95,4.60,mwr,NaN,34.0,1.8970,0.99,...,2020-11-24T01:20:30.040Z,"48 km WNW of Abra Pampa, Argentina",earthquake,7.60,6.20,NaN,NaN,reviewed,us,guc
89,2020-11-24T03:49:00.985Z,17.3788,-93.9465,175.42,4.50,mb,NaN,132.0,2.0470,0.63,...,2020-11-24T05:38:28.040Z,"33 km NE of La Horqueta (Poblado Doce), Mexico",earthquake,9.90,10.50,0.064,71.0,reviewed,us,us
108,2020-11-24T00:32:03.180Z,18.5315,-68.5501,151.00,3.57,md,13.0,193.0,0.1696,0.05,...,2020-11-24T01:25:34.040Z,"16 km WSW of Punta Cana, Dominican Republic",earthquake,8.75,0.84,0.250,12.0,reviewed,pr,pr
87,2020-11-24T03:51:05.311Z,-22.4597,-68.6890,113.62,4.20,mb,NaN,38.0,0.6790,0.80,...,2020-11-24T05:31:46.040Z,"24 km E of Calama, Chile",earthquake,4.50,6.00,0.153,12.0,reviewed,us,us
100,2020-11-24T01:44:43.481Z,-33.6955,-70.3317,111.79,4.60,mb,NaN,69.0,0.0610,0.41,...,2020-11-24T12:03:48.630Z,"Region Metropolitana, Chile",earthquake,4.60,4.60,0.183,9.0,reviewed,us,us
212,2020-11-23T15:25:22.177Z,51.7129,179.6182,110.81,3.00,ml,NaN,343.0,0.9850,0.42,...,2020-11-23T23:27:47.688Z,"259 km W of Adak, Alaska",earthquake,20.30,12.60,0.148,6.0,reviewed,us,us


You can also find ```max``` and ```min``` values in a column. Eg:

In [22]:
try:
    print(QUAKE_DF["depth"].max())
except:
    print("Probably QUAKE_DF not correctly set")    

255.95


In [23]:
try:
    print(QUAKE_DF["depth"].min())
except:
    print("Probably QUAKE_DF not correctly set")    

-1.64999997615814


### Question 2a: Find Powerful Quakes

Write a function `powerful_quakes` that takes a numerical argument and returns a `DataFrame` including
all the quakes in `QUAKE_DF` that have a magnitude greater than or equal to the given argument.

In [24]:
# Complete question 2a answer cell

def powerful_quakes(mag):
    ## This is just returning an empty DataFrame you need to code it to return
    ## a DataFrame with all quakes of magnitude greater than or equal to mag
    
    df = QUAKE_DF.copy()
    
    return df[df['mag'] >= mag]

In [ ]:
#Please do not run this cell
do_tests( powerful_quakes )

### Question 2b: Find `n+` most powerful earthquakes

Produce a `DataFrame` with rows that represent the `n`(or maybe more) most powerful quakes
in descending order of magnitude. The returned ``DataFrame`` should show at least `n`
quakes and may sometimes show more since we do not want to leave out any quake that is equally
powerful as the last quake listed in the `DataFrame`.
More specificially, we want the function to return a `DataFrame` that:
* has exactly the same column names as `QUAKES_DF`,
* each row has the same vaues in each column as a corresponding row in `QUAKES_DF` (but it does not matter whether the row indices in the returned `DataFrame` are the same or different from corresponding rows in `QUAKES_DF`),
* the rows of the returned `DataFrame` are ordered in _descending order_ of their magnitude column value,
* contains all and only those rows `QUAKES_DF`, such that there are fewer than `n` other rows in
  `QUAKES_DF` that have a higher magnitude.


#### Note:
The above definition of the requirements is clear and precise. Though you may ask for help and advice regarding implementation, you will not be given help with understanding the specification. 

In [25]:
# Question 2b answer cell

def most_powerful_n_quakes(n):
    
    df = QUAKE_DF.copy()
    df.sort_values(by='mag', ascending=False, inplace=True)
    
    min_mag = df.iloc[n-1]['mag']
        
    return df[df['mag']>=min_mag] 
    # Edit this function to make it return a DataFrame of 
    # the 'top n' quakes of the all_day.csv file

In [ ]:
#Please do not run this cell
do_tests( most_powerful_n_quakes )

### Distance between locations on the Earth's surface

Clearly, when dealing with data pertaining to locations in space, the distance between such locations is often of great significance when interpreting or extracting further information from the data.

To help answer the following questions you are provided with the function ```haversine_distance```, 
which implements the [_Haversine formula_](https://en.wikipedia.org/wiki/Haversine_formula) to find the surface distance in kilometres between two locations, that are specified in terms of
latitude and longitude values. When finding distances betwen points on the surface of the
Earth we need to use this formula, rather than the simpler Pythagorean distance formula,
because the Earth's surface is a sphere.

In [26]:
## Function to compute distance between locations (kilometres) 
# Returns the surface distance in meters, according to the Haversine formula,
# between two locations given as (latitude, longitude) coordinate pairs.

import math
def haversine_distance( loc1 , loc2 ): 
    '''finds the distance (m) between 2 locations, where locations are defined by
    longitudes and latitudes'''
    lat1, lon1 = loc1
    lat2, lon2 = loc2
    radius = 6371  # kilometers
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) * math.sin(dlat / 2) +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) * math.sin(dlon / 2))
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    d = radius * c
    return d

### Question 2c: Sort quakes by distance from a given location

Write a function `quake_distance_from_loc_dataframe(loc)` satisfying the following requirements:
* It takes a location argument `(latitude,longitude)` consisting of a pair of `float`s 
    (Note: this is a single argument but consists of a pair of values represented as Python `tuple`.)

* It returns a `DataFrame` object derived from `QUAKE_DF` but with one extra column `distance_from_loc`, giving the distance of each quake from the given location. 

* The rows of the returned `DataFrame` should be the same as those in `QUAKE_DF` except for the aditional `distance_from_loc` column. (However, it is not necessary to preserve the index of the `DataFrame`, this will be ignored when your solution is tested.)

* The rows of the returned `DataFrame` should be _sorted_ in order of _increasing_ values of `distance_from_loc`. 

* The original DataFrame `QUAKE_DF` should not be altered by the execution of 
`quake_distance_from_loc_dataframe(loc)`.

#### Note:
You will need to do some research to find out how to create a new column and set its values.

In [27]:
def quake_distance_from_loc_dataframe(loc):
    ## Replace with code so that the function returns a DataFrame in accord with the
    ## specification given above
    df = QUAKE_DF.copy()
    df['distance_from_loc'] = df.apply(lambda x: haversine_distance(loc, (x['latitude'], x['longitude'])), axis=1)
    df.sort_values(by='distance_from_loc', ascending=True, inplace=True)
    
    return df    

In [ ]:
#Please do not run this cell
do_tests( quake_distance_from_loc_dataframe )

### Question 2d: Identifying Endangered Cities

The idea of this question is to identify possible emergency situations by finding
cities that are likely to suffer from the effects of
an earthquake.

#### Effect of an qarthquake at a distance from its epicenter
The effect of an earthquake on a city or person will depend on their distance from the source of the quake. The effect of an earthquake will depend on many factors and even the dependence on distance to source is very complex. However, after a bit of background research, Brandon has come up with a simple formula which hopefully at least gives a very crude estimate of relative effect of a quake with a particular magnitude and depth on a surface location at a known surface distance from the quake's epicenter. The calculated effective magnitude of an earthquake will be less that the source magnitude, for instance a magnitude 9 quake at a depth of 100km (which is likely to be extremely destructive, would have an effective magnitude of 5 at its epicentre (directly above the source) and 3.585 at a point on the earth surface 500km away from the epicenter.

In [28]:
def effective_magnitude( magnitude, depth, surface_distance ):
    energy = 10**magnitude  # convert logarithmic magnitude to a linear energy value
    if depth < 1:   # Crude fix for small or negative depths (can occur where land is above sea level)
        depth = 1
    ## Calculate distance to source by Pythagorus (ignoring curvature of surface)
    dist_to_source_squared =  depth**2 + surface_distance**2
    ## Apply inverse square distance multiplier to get energy density at distance from source
    ## (Ignores damping effects)
    attenuated_energy = energy/dist_to_source_squared
    attenuated_magnitude =  math.log10(attenuated_energy) ## Convert back to a log base 10 scale
    return attenuated_magnitude

# Some test cases.
#effective_magnitude(9,100,500)
#effective_magnitude(6,50, 100)

The _epicenter_ of an earthquake is the point on the earth that is directly above its source. Thus the effective magnitude at the epicenter is just the effective magnitude of the quake at surface distance zero:

In [29]:
def epicenter_magnitude( magnitude, depth ):
    return effective_magnitude( magnitude, depth, 0)

**Note:** For any given quake, its `effective_magnitude` at any point on Earth is always less than or equal to its `epicenter_magnitude`.

### Specification of the `endangered_cities` function

Now we get to the specification of the function.

Write a function `endangered_cities( minimum_population, minimum_effective_magnitude)`
that takes 
two numerical arguments: an `int` (`minimum_population`) and a `float` (`minimum_effective_magnitude`) 
and  returns a `list` specifying all those cities listed in 
the `WC_DF` such that:
* the city has  a population greater than or equal to the given `minimum_population`;

* for at least one of the quakes recorded in `QUAKE_DF`, the `effective_magnitude` (as determined by the
  function defined above) of that quake
   at the location of the city (as given in `worldcities.csv`) is equal to or greater than the  `minimum_effective_magnitude`.
   
* each city in the list should be represented by a tuple `(city_ascii, country, (lat, lng))`, giving
the city name in ASCII form, the country and the location (as a latitude, longitude pair).

* the returned list should be ordered aphabetically, primarily in terms of `country` and secondarily
  cities of the same country should be ordered in terms of `city_ascii`. 
  This ordering is illustrated by the sample output given below.

* A final condition is that the function should run in a reasonable time frame such that it can be expected to give correct and complete results after no more that 5 minutes excecution time. To ensure that this is feasible, the example cases you will be tested on, will be ones for which Brandon's solution ran in less than 2 minues on the Gradescope autograder platform. (All submissions will be tested on the same saved copy of `all_day.csv` previously downloaded from the USGS feed.)
   
##### Example Output:   

<pre>
 in [165]    %time        
             endangered_cities(200000, 0.5)
 
 out[165]    CPU times: user 1min 58s, sys: 768 µs, total: 1min 58s
             Wall time: 1min 58s
             [('Baghlan', 'Afghanistan', (36.1393, 68.6993)),
              ('Kunduz', 'Afghanistan', (36.728, 68.8725)),
              ('Mazar-e Sharif', 'Afghanistan', (36.7, 67.1)),
              ('Ambon', 'Indonesia', (-3.7167, 128.2)),
              ('Denov', 'Uzbekistan', (38.2772, 67.8872))]
</pre>

##### Notes:

* Your code should make use of the functions defined above: `haversine_distance`, `effective_magnitude` and `epicenter_magnitude`. You are advised not to change these otherwise you may get different results from what the Autograder is expecting.

* The calculation of this function is computationally intensive. For prelimiary testing you might use smaller data sets by, say only using cities in one country. 

* There are many optimisations that can be done to reduce the computational cost of finding the endangered cities. For example, calculating `epicenter_magnitude` enables weaker earthquakes to be identified withought the need to determine their distance from every city on earth. It may also be useful to know that when calcuating the effective magnitude of a given quake (with a specific depth) at different locations, the value always decreases as surface distance from the quake increases. 

* Remember that the autograder will run your actual code, so will probably take a while to grade this function. However, it only runs the function once and then performs 3 different tests on the value that are returned.

* The autograder will run on a presaved set of quakes (so the test is the same for all submissions). However, your code should run on any version of `all_day.csv` that you download from the USGS live feed.

In [181]:
## 2d Answer Code Cell
def endangered_cities(min_population, min_effective_magnitude):
    
    df_wc = WC_DF.dropna(subset='population', inplace=False)
    df_qk = QUAKE_DF.copy()
    list_cities = []
    
#     test_cities = ['Nuku`alofa']
#     df_wc = df_wc[df_wc['city_ascii'].isin(test_cities)]
#     test_countries = ['Japan', 'Indonesia']
#     df_wc = df_wc[df_wc['country'].isin(test_countries)]
    
    df_wc = df_wc[df_wc['population']>=min_population]
    df_wc['lat_lng'] = df_wc.apply(lambda x: (x['lat'], x['lng']), axis=1)
    
    df_qk['epicentre_mag'] = df_qk.apply(lambda x: epicenter_magnitude(x['mag'], x['depth']), axis=1)
    df_qk = df_qk[df_qk['epicentre_mag']>=min_effective_magnitude]
    df_qk['lat_lng'] = df_qk.apply(lambda x: (x['latitude'], x['longitude']), axis=1)
    
    for i,city in df_wc.iterrows():
#         print(city['city'])
        df_qk['city_dist'] = df_qk.apply(lambda x: haversine_distance(x['lat_lng'],city['lat_lng']), axis=1)
        df_qk.sort_values(by='city_dist', ascending=True, inplace=True)
        
#         print(df_qk.head())
#         print(df_qk.mag.iat[0])
        max_eff_mag = effective_magnitude(df_qk.mag.iat[0], df_qk.depth.iat[0], df_qk.city_dist.iat[0])
        if max_eff_mag >= min_effective_magnitude:
            list_cities.append((city['city_ascii'], city['country'], city['lat_lng'], max_eff_mag))
    
    return list_cities

In [183]:
my_list = endangered_cities(20000, 0.5)
print(my_list)
print(len(my_list))

[('Nuku`alofa', 'Tonga', (-21.1385, -175.2206), 0.895262639828624), ('Ambon', 'Indonesia', (-3.7167, 128.2), 0.5451497530652795), ('Maina', 'Guam', (13.4692, 144.7332), 1.2612861770731074), ('Amahai', 'Indonesia', (-3.3281, 128.9405), 0.5906914756662714)]
4


In [200]:
df_quake = QUAKE_DF.copy()
df_cities = WC_DF.copy().dropna()

df_quake['lat_lng'] = df_quake.apply(lambda x: (x['latitude'], x['longitude']), axis=1)
df_cities['lat_lng'] = df_cities.apply(lambda x: (x['lat'], x['lng']), axis=1)
df_quake['closest_city'] = 0
df_quake['city_dist'] = 0
df_quake['eff_mag'] = 0

for i,quake in df_quake.iterrows():
    df_cities['quake_dist'] = df_cities.apply(lambda x: haversine_distance(x['lat_lng'], quake['lat_lng']), axis=1)
    df_cities.sort_values(by='quake_dist', ascending=True, inplace=True)
    df_quake.closest_city.at[i] = df_cities.city_ascii.iat[0]
    df_quake.city_dist.at[i] = df_cities.quake_dist.iat[0]
#     print(quake['mag'])
#     print(quake['depth'])
#     print(quake['city_dist'])
    df_quake.eff_mag.at[i] = effective_magnitude(quake['mag'], quake['depth'], df_quake.city_dist.at[i])

df_quake


,time,latitude,longitude,depth,mag,magType,nst,gap,dmin,rms,...,depthError,magError,magNst,status,locationSource,magSource,lat_lng,closest_city,city_dist,eff_mag
0,2020-11-24T14:32:03.660Z,38.093800,-117.521200,6.20,2.20,ml,12.0,301.36,0.196000,0.08,...,10.70,NaN,NaN,automatic,nn,nn,"(38.0938, -117.5212)",Carson City,226.318355,-2.509765
1,2020-11-24T14:29:48.800Z,38.803001,-122.772331,0.76,0.40,md,10.0,76.00,0.010140,0.03,...,0.72,0.160,2.0,automatic,nc,nc,"(38.8030014, -122.7723312)",Sacramento,116.196306,-3.730417
2,2020-11-24T14:29:38.517Z,60.563200,-151.575600,58.20,2.00,ml,NaN,NaN,NaN,0.69,...,1.10,NaN,NaN,automatic,ak,ak,"(60.5632, -151.5756)",Whitehorse,898.734270,-3.909080
3,2020-11-24T14:28:38.100Z,38.175000,-117.872200,6.60,1.20,ml,13.0,108.70,0.015000,0.09,...,0.80,NaN,NaN,automatic,nn,nn,"(38.175, -117.8722)",Carson City,195.667754,-3.383532
4,2020-11-24T14:07:34.177Z,-22.197700,-176.098100,57.45,5.30,mb,NaN,83.00,6.560000,0.80,...,6.00,0.065,78.0,reviewed,us,us,"(-22.1977, -176.0981)",Nuku`alofa,148.640044,0.895263
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
212,2020-11-23T15:25:22.177Z,51.712900,179.618200,110.81,3.00,ml,NaN,343.00,0.985000,0.42,...,12.60,0.148,6.0,reviewed,us,us,"(51.7129, 179.6182)",Petropavlovsk-Kamchatskiy,1427.508251,-3.311766
213,2020-11-23T15:22:37.770Z,38.823502,-122.808998,2.75,0.55,md,10.0,92.00,0.000847,0.01,...,1.05,NaN,1.0,automatic,nc,nc,"(38.8235016, -122.8089981)",Sacramento,119.805038,-3.607179
214,2020-11-23T15:13:19.000Z,17.949800,-66.931000,11.00,2.13,md,7.0,212.00,0.057600,0.17,...,0.87,0.190,5.0,reviewed,pr,pr,"(17.9498, -66.931)",Higuey,201.711507,-2.480751
215,2020-11-23T15:11:43.540Z,32.748833,-115.814000,5.37,1.74,ml,19.0,65.00,0.123700,0.23,...,1.11,0.159,29.0,reviewed,ci,ci,"(32.7488333, -115.814)",Mexicali,33.129045,-1.311681


In [35]:
#Please do not run this cell
do_tests( endangered_cities )

NameError: name 'do_tests' is not defined

## Optional Exercises

Having got this far, you may find it interesting and informative to do some more processing
of the city and earthquake information.
Since the previous exercises were designed so they can be quickly and reliably assessed by the autograding software, they involve coding particular functions with very specific requirements. But the following exercises are more open ended and give suggestions for interactive and visual use of the city and earthquake data.

To ensure that your assignment submission works with the autograding software when submitted, it is recommended that you now save this file and make a new copy with a different name, such as `Earthquakes_optional.ipynb`. Then use the new file to continue with the optional exercises.

### Constructing a city risk status alert `DataFrame`

A government or other organisation may want to monitor a certain list of cities with regard to whether
they may be at risk of earthquake damage. To answer this question you should create a function
that uses the `endangered_cities` function you have defined above to create such a `DataFrame`.

Your function `city_risk_alert` should return a pandas DataFrame that includes the status of ```'ENDANGERED'``` or ```'SAFE'``` for a certain city. The dataframe should also contain the city name, country and status for each city input. You could also extend this to add more columns showing
things like the distance and magnitude of the nearest earthquake. And you could perhaps make it
so any endangered cities were put at the top of the list.

For example:
```
display( city_risk_alert( ['Rome', 'Milan', 'Pisa'] )
```

might give the following output:

 
| city  | country|status|
|-------|-------|-----|
| Pisa  | Italy | ENDANGERED |
| Rome  | Italy | SAFE |
| Milan | Italy | SAFE |
 

## Visualisation Exercise: display endangered cities on a map

The code below creates a Map object using the ```ipyleaflet``` module and uses this to
display powerful quakes on the map. If you have coded the `powerful_quakes` function for
Question 2b above, the code in the cell below the map should draw the detected powerful
quakes onto the map at their correct locations.

To install the ```ipyleaflet``` module use ```pip3 install ipyleaflet```. If the map does not display after installation be sure to restart the kernel, and close and reopen this file. We provide the ```draw_circle_on_map``` function, this adds circles to a specified location on the map, where the location is defined by longitudes and latitudes.

In [ ]:
from ipyleaflet import Map, basemaps, basemap_to_tiles, Circle, Polyline
from ipywidgets import Layout

LEEDS_LOC  = ( 53.8008,  -1.5491  ) # Here we define the longitude and latitude of Leeds
WORLD_MAP = Map(basemap=basemaps.OpenTopoMap, center=LEEDS_LOC, zoom=1.5,
                layout=Layout(height="500px")) # Here we create a map object centred on Leeds

WORLD_MAP

In [ ]:
def draw_circle_on_map( a_map, location, radius = 1000, color="red", fill_color=None ):
    if not fill_color:
        fill_color = color
    circle = Circle()
    circle.location = location
    circle.radius = radius
    circle.color = color
    circle.fill_color = fill_color
    a_map.add_layer(circle)

# This will edit your previous map rather than produce a new one    
draw_circle_on_map(WORLD_MAP, LEEDS_LOC, color="green" ) 

def display_powerful_quakes_on_map(mag):
    powerful = powerful_quakes(3)
    for i, quake in powerful.iterrows():
        draw_circle_on_map( WORLD_MAP,
                            (quake["latitude"],quake["longitude"]), 
                            radius= 20000*int(quake["mag"]) )

display_powerful_quakes_on_map(3)

### More Ideas for Graphical Display
It would be nice to also see the endangered cities on the map. For an ambitious exercise you
could see if you can draw lines on the map running from the locations of powerful
earthquakes to the cities that are endangered by them.